### PR TITLE
Alert queueing

### PR DESCRIPTION
**Pull request by jsom**

### CODE REVIEW

File: ExampleApp/ExampleViewController.swift [hunk 1]
```diff
@@ -26,7 +27,7 @@ class ExampleViewController: UIViewController, UITableViewDataSource, UITableVie
         super.viewDidLoad()
 
         self.tradeItLauncher = TradeItLauncher(apiKey: API_KEY, environment: ENVIRONMENT)
-        self.alertManager = TradeItAlertManager()
+        self.alertManager = TradeItAlertManager.sharedInstance
```
Comment (jsom) [inline ExampleApp/ExampleViewController.swift]: Now a singleton

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 2]
```diff
@@ -63,14 +69,23 @@ import UIKit
...
-        viewController.presentViewController(alertController, animated: true, completion: nil)
+        add(alertClosure: {
+            let alert = TradeItAlertProvider.provideSecurityQuestionAlertWith(
+                alertTitle: "Security Question",
+                alertMessage: securityQuestion.securityQuestion ?? "No security question provided.",
+                multipleOptions: securityQuestion.securityQuestionOptions ?? [],
+                alertActionTitle: "Submit",
+                onAnswerSecurityQuestion: { answer in
+                    onAnswerSecurityQuestion(withAnswer: answer)
+                    self.activeAlertFinished()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: This is pretty funky. I'm not totally happy with having to remember to call `activeAlertFinished`. Perhaps the caching is better in the provider?

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 3]
```diff
@@ -79,11 +94,41 @@ import UIKit
...
+                onCanceledActionTapped: {
+                    onCancelActionTapped?()
+                    self.activeAlertFinished()
+                }
+            )
+            viewController.presentViewController(alert, animated: true, completion: nil)
+        })
+    }
+
+    func add(alertClosure alertClosure: () -> Void) {
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: Not sure `alertClosure` is the best name for this. Generator? Context?

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 4]
```diff
@@ -1,9 +1,15 @@
 import UIKit
 
 @objc public class TradeItAlertManager: NSObject {
+    public static let sharedInstance = TradeItAlertManager()
+
+    var alertClosureQueue: [() -> Void] = []
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: We store closures that will generate the alert with the correct viewController etc that it was originally called with.

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 5]
```diff
@@ -79,11 +94,41 @@ import UIKit
...
+    func displayNextAlert() {
+        if alertClosureQueue.count <= 0 { return }
+        let alertClosure = alertClosureQueue.removeFirst()
+        alertBeingPresented = true
+        alertClosure()
+    }
+
+    func activeAlertFinished() {
+        alertBeingPresented = false
+        displayNextAlert()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: This is starting to get messy. I want to make a new object and add some unit tests.

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 6]
```diff
@@ -79,11 +89,48 @@ import UIKit
...
+            viewController.presentViewController(alert, animated: true, completion: nil)
+        })
+    }
+
+}
+
+class TradeItAlertQueue {
+    static let sharedInstance = TradeItAlertQueue()
+
+    var alertClosureQueue: [() -> Void] = []
```
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: Maybe use tuple here?
```
var alertQueue = [(UIViewController, UIAlertController)]()
```

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 7]
```diff
@@ -79,11 +89,48 @@ import UIKit
...
+    var alreadyPresentingAlert = false
+
+    private init() {}
+
+    func add(alertClosure alertClosure: () -> Void) {
+        alertClosureQueue.append(alertClosure)
+        self.showNextAlert()
+    }
+
+    func showNextAlert() {
```
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: private

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 8]
```diff
@@ -79,11 +89,48 @@ import UIKit
...
+                }
+            )
+            viewController.presentViewController(alert, animated: true, completion: nil)
+        })
+    }
+
+}
+
+class TradeItAlertQueue {
+    static let sharedInstance = TradeItAlertQueue()
```
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: private

File: TradeItIosTicketSDK2/TradeItAlertManager.swift [hunk 9]
```diff
@@ -79,11 +89,48 @@ import UIKit
...
+
+    private init() {}
+
+    func add(alertClosure alertClosure: () -> Void) {
+        alertClosureQueue.append(alertClosure)
+        self.showNextAlert()
+    }
+
+    func showNextAlert() {
+        if alreadyPresentingAlert || alertClosureQueue.count <= 0 { return }
```
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItAlertManager.swift]: `isEmpty`

**Pull request merged**